<a href="https://colab.research.google.com/github/dscohen75/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [1]:
# YOUR WORK HERE

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-08-11 19:10:10--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.1s    

2020-08-11 19:10:11 (154 KB/s) - ‘student.zip’ saved [20478/20478]



In [2]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [5]:
import pandas as pd
import numpy as np
from scipy import stats

In [6]:
df = pd.read_csv('student-mat.csv', sep=';')

In [7]:
df.head()

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

In [8]:
df.shape

(395, 33)

In [9]:
df.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

In [10]:
pd.crosstab(df['school'], df['sex'], margins=True)

sex       F    M  All
school               
GP      183  166  349
MS       25   21   46
All     208  187  395

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


Let's look at travel time and address:

In [14]:
# YOUR WORK HERE

contingency = pd.crosstab(df['traveltime'], df['address'])

contingency

address      R    U
traveltime         
1           35  222
2           34   73
3           14    9
4            5    3

Intuitively I would expect that address and travel to school would be related, so that's worth testing.

H0: Address and travel time to school are not related.

H1: Address and travel time to school are related.

In [17]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 43.96115908406922
p value: 1.5381274806593009e-09
dof: 3
expected_frequencies: 
 [[ 57.2556962  199.7443038 ]
 [ 23.83797468  83.16202532]
 [  5.12405063  17.87594937]
 [  1.78227848   6.21772152]]


Indeed, with a Chi squared statistic of 43.96 and a p-value of .0000000015, we can confidently say that address and travel time to school are strongly related.


Let's try to find a pair of variables that we'd really assume are independent.

Looking at mother's job versus child's sex:

H0: Mother's job is unrelated to the child's sex.

H1: The mother's job depends on the child's sex.

In [27]:
contingency2 = pd.crosstab(df['sex'], df['Mjob'])

contingency2 

Mjob  at_home  health  other  services  teacher
sex                                            
F          42      19     74        54       19
M          17      15     67        49       39

In [28]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency2)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 17.483557532214064
p value: 0.0015564385627536573
dof: 4
expected_frequencies: 
 [[31.06835443 17.90379747 74.24810127 54.23797468 30.54177215]
 [27.93164557 16.09620253 66.75189873 48.76202532 27.45822785]]


With a Chi-squared statistic of 17.48 and p-value of 0.01, we reject the null hypothesis that mother's job and child's sex are unrelated. 

Let's try family size and study time.

H0: Study time is unrelated to family size.

H1: Study time depends upon family size.

In [25]:
contingency3 = pd.crosstab(df['famsize'], df['studytime'])

contingency3

studytime   1    2   3   4
famsize                   
GT3        74  131  57  19
LE3        31   67   8   8

In [26]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency3)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 11.09435481782051
p value: 0.011226452773765971
dof: 3
expected_frequencies: 
 [[ 74.69620253 140.8556962   46.24050633  19.20759494]
 [ 30.30379747  57.1443038   18.75949367   7.79240506]]


With a Chi-squared statistic of 11.09 and a p-value of .01, we reject the null hypothesis that study time is independent of family size.

Lot's of variables in this dataset seem to be dependent upon each other.
Let's try guardian and sex.

H0: Who the child's guardian is does not depend upon the child's sex.
H0: Who the chid's guardian is depends on the child's sex.

In [30]:
contingency4 = pd.crosstab(df['guardian'], df['sex'])

contingency4

sex         F    M
guardian          
father     43   47
mother    145  128
other      20   12

In [32]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency4)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 2.125939043359857
p value: 0.3454285283131414
dof: 2
expected_frequencies: 
 [[ 47.39240506  42.60759494]
 [143.75696203 129.24303797]
 [ 16.85063291  15.14936709]]


Finally we've found a pair of variables that are not dependent on each other! With a Chi-squared statistic of 2.12 and p-value of 0.35, we do not reject the null hypothesis that guardian and sex are independent of each other.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [34]:
# YOUR WORK HERE

observed = pd.crosstab(df['guardian'], df['sex'], margins=True)

observed

sex         F    M  All
guardian               
father     43   47   90
mother    145  128  273
other      20   12   32
All       208  187  395

In [49]:
# row totals as a numpy array

row_totals = df['guardian'].value_counts().sort_index().values

row_totals

array([ 90, 273,  32])

In [50]:
# column totals as a numpy array

col_totals = df['sex'].value_counts().sort_index().values

col_totals

array([208, 187])

In [39]:
sample_size = df.shape[0]

sample_size

395

In [51]:
expected = np.array([[0.0, 0.0],
                    [0.0, 0.0],
                    [0.0, 0.0]])

for row_index, row in enumerate(row_totals):
  for column_index, col in enumerate(col_totals):
    expected_value = (row*col)/sample_size
    expected[row_index][column_index] = expected_value

expected


array([[ 47.39240506,  42.60759494],
       [143.75696203, 129.24303797],
       [ 16.85063291,  15.14936709]])

In [56]:
# Need to reset observed to be the crosstab without the margins, 
# in order to use it for calculating the Chi2 statistic.

observed = pd.crosstab(df['guardian'], df['sex'])

observed

sex         F    M
guardian          
father     43   47
mother    145  128
other      20   12

In [57]:
type(observed)

pandas.core.frame.DataFrame

In [58]:
# Also need to take .values, and turn this into a numpy array
observed = observed.values

observed

array([[ 43,  47],
       [145, 128],
       [ 20,  12]])

In [59]:
type(observed)

numpy.ndarray

In [55]:
chi2 = ((observed - expected)**2 / expected).sum()

chi2

2.125939043359857

This is the same result that the stats.chi2_contingency function gave us.

This Chi2 statistic has 
dof = (Number of rows - 1) * (Number of columns - 1)
= 2 * 1 = 2

Using the applet at https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html
we get a p-value of 0.35, which also matches scipy's result.

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)